File: gad-rust/src/uuu/parse/lex.rs

// lex.rs
//
pub unsafe fn lexer(pp :&str) -> Vec<String> {
  let mut mm :String = pp.to_string();
  let baby = mm.as_mut_vec();
  let mut ret :Vec<String> = Vec::new();
  let n = baby.len();
  let mut i = 0;
  let mut buf :Vec<u8> = Vec::new();
  let mut blank :bool = false;
  let mut quot :bool = false;
  while i < n {
    let mut a = baby[i];
    if a == b'\r' { a = b' '; };
    if a == b'\n' { a = b' '; };
    if a == b'\"' {
      if quot {
        blank = false; quot = false;
      } else {
        buf.push(a);
        quot = true; blank = false; 
      };
      i += 1; continue;
    };
    if a == b' ' {
      if quot {
        buf.push(a);
        i += 1; continue;
      };
      if blank {
        i += 1; continue;
      };
      if buf.len() > 0 {
        blank = false;
        ret.push(String::from_utf8_unchecked(buf));
        buf = Vec::new();
      };
      i += 1; 
      continue;
    };
    buf.push(a);
    i += 1;
  };
  if buf.len() > 0 {
    ret.push(String::from_utf8_unchecked(buf));
  };
  // output
  let np = ret.len();
  println!("!parsed {}",np);
  i = 0;
  while i < np {
    let be = &ret[i]; 
    print!("<! { } !>",be);
    i += 1;
  };
  print!("\n");
  return ret;
}

File: lex.go

// lex.go

package main

import "fmt"
import "strings"

func Lexer(pp string) ( ret [256]string , nret int  ) {
  var baby string = pp;
  ret[0]=""
  var n = len(baby);
  var i = 0;
  var buf string = "";
  var blank bool = false;
  var quot bool = false;
  for i < n {
    var a = baby[i:i+1];
    if a == "\r" { a = " "; };
    if a == "\n" { a = " "; };
    if a == "\"" {
      if quot {
        blank = false; quot = false;
      } else {
        buf = strings.Join( []string { buf, a } , "" )
        quot = true; blank = false; 
      };
      i += 1; continue;
    };
    if a == " " {
      if quot {
        buf = strings.Join( []string { buf, a },"");
        i += 1; continue;
      };
      if blank {
        i += 1; continue;
      };
      if len(buf) > 0 {
        blank = false;
        ret[nret] = buf; nret += 1; buf = "";
      };
      i += 1; 
      continue;
    };
    buf = strings.Join([]string{ buf ,a },"");
    i += 1;
  };
  if len(buf) > 0 {
    ret[nret] = buf;
    nret += 1;
  };
  // output
  i = 0;
  for i < nret {
    var be = ret[i]; 
    fmt.Print("<! {")
    fmt.Print(be)
    fmt.Print("} !>");
    i += 1;
  };
  fmt.Println("\n");
  return;
}